src/Provider/TwigUsageProvider.php

```diff
@@ -0,0 +1,198 @@
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\BetterReflection\Reflection\Adapter\ReflectionMethod;
+use PHPStan\Node\InClassNode;
+use PHPStan\Reflection\ExtendedMethodReflection;
+use ShipMonk\PHPStan\DeadCode\Graph\ClassMethodRef;
+use ShipMonk\PHPStan\DeadCode\Graph\ClassMethodUsage;
+use ShipMonk\PHPStan\DeadCode\Graph\UsageOrigin;
+use function count;
+use function explode;
+use function in_array;
+
+class TwigUsageProvider implements MemberUsageProvider
+{
+
+    private bool $enabled;
+
+    public function __construct(?bool $enabled)
+    {
+        $this->enabled = $enabled ?? $this->isTwigInstalled();
+    }
+
+    private function isTwigInstalled(): bool
+    {
+        return InstalledVersions::isInstalled('twig/twig');
+    }
+
+    public function getUsages(Node $node, Scope $scope): array
+    {
+        if (!$this->enabled) {
+            return [];
+        }
+
+        $usages = [];
+
+        if ($node instanceof InClassNode) { // @phpstan-ignore phpstanApi.instanceofAssumption
+            $usages = [
+                ...$usages,
+                ...$this->getMethodUsagesFromReflection($node),
+            ];
+        }
+
+        if ($node instanceof New_) {
+            $usages = [
+                ...$usages,
+                ...$this->getMethodUsageFromNew($node, $scope),
+            ];
+        }
+
+        return $usages;
+    }
+
+    /**
+     * @return list<ClassMethodUsage>
+     */
+    private function getMethodUsageFromNew(New_ $node, Scope $scope): array
+    {
+        if (!$node->class instanceof Name) {
+            return [];
+        }
+
+        $className = $node->class->toString();
+
+        if (!in_array($className, [
+            'Twig\TwigFilter',
+            'Twig\TwigFunction',
+            'Twig\TwigTest',
+        ], true)) {
+            return [];
+        }
+
+        $arg = $node->getArgs()[1] ?? null;
+
+        if ($arg === null) {
+            return [];
+        }
+
+        $argType = $scope->getType($arg->value);
+
+        $callable = [];
+
+        if ($argType->isConstantArray()->yes()) {
+            $arrayType = $argType->getConstantArrays()[0] ?? null;
+
+            if ($arrayType !== null) {
+                foreach ($arrayType->getValueTypes() as $valueType) {
+                    $stringType = $valueType->getConstantStrings()[0] ?? null;
+
+                    if ($stringType !== null) {
+                        $callable[] = $stringType->getValue();
+                    }
+                }
+            }
+        } else {
+            $stringType = $argType->getConstantStrings()[0] ?? null;
+
+            if ($stringType !== null) {
+                $callable = explode('::', $stringType->getValue());
+            }
+        }
+
+        if (count($callable) === 2) {
+            $usage = new ClassMethodUsage(
+                UsageOrigin::createVirtual($this, VirtualUsageData::withNote('Used in ' . $className)),
```

Comment:
For AST based detections, `::createRegular` should be used, this enables proper transitivity detection.